sensory_jamming.py

```diff
@@ -0,0 +1,342 @@
+        
+    def Identification(self):
+        
+        for agent in range(self.popsize):
+            # for each agent in the bat population:
+            self.all_ID[int(agent)] = agent
+                # fill-in a list of all agents ID 
+                # from 0 to n = (population size - 1)
+            
+    def Positions(self):
+        
+        for ID in self.all_ID:
+            # for each agent in the bat population:
+           self.agent_initx = rd.uniform(0, self.boxsize[0])
+               # give each agent a random coordinate on the x-axis, 
+               # within the environment boundaries
+           self.agent_inity = rd.uniform(0, self.boxsize[1])
+               # give each agent a random coordinate on the y-axis, 
+               # within the environment boundaries
+           self.all_initpos[int(ID),0] = self.agent_initx
+               # store the x-coordinate in an array
+           self.all_initpos[int(ID),1] = self.agent_inity
+               # store the y-coordinate in an array
+    
+    def Timeline(self):
+            
+        for iteration in range(self.simduration):
+            # for each time step in the simulation:
+            self.realtime += float(self.d_t)
+                # increment timecount with time resolution. 
+            self.timeclock[int(iteration)] = self.realtime
+                # store the new, incremented time count in timeclock.
+
+    class Bat_Jamming_00:
+        # Class implementation nested into Launcher class implementation.
+        # Sets an agent's movement, calling and hearing rules.
+    
+        def __init__(self, ID, movdirection, flightspeed, IPI, max_hear_dist):
+            self.ID = int(ID)
+                # integer. Identification number of the focal agent.
+            self.boxsize = env.boxsize
+                # takes boxsize from env (class Launcher).
+            self.d_t = env.d_t
+                # takes d_t from env (class Launcher).
+            self.movdirection = float(movdirection)
+                # float. Direction (in radians) towards which the agent flies.
+                # Must be between -pi and pi (asserted later).
+            self.stepsize = float(flightspeed * self.d_t) 
+                # float. Distance in meters covered by the agent in 1 time step
+            self.IPI = int(float(IPI * m.pow(10,-3)) / env.d_t)
+                # int. Agent's inter-pulse interval, converted in time steps.
+            self.speedsound = 340.29 
+                # float. Speed of sound at sea level in m/s.
+            self.max_timestore = float(max_hear_dist / self.speedsound)
+                # float. Time maximal, in seconds, during which a sound can 
+                # travel, before its intensity passes below the hearing 
+                # threshold of the agent.
+            self.ring_width = float(1) 
+                # float. Spatial difference between start and end of call (m).
```

Comment:
'spatial difference' feels vague - perhaps we want to say - the ring_width is the difference in radii of the two concentric circles ( in 2D) which form the start and the end of the bat call.